File: leetcode/python/46_permutations.py
class Solution:
    def permute(self, nums: list[int]) -> list[list[int]]:
        res: list[list[int]] = []
        self.helper(set[int](nums), [], res)
        return res
    
    def helper(self, nums: set[int], group: list[int], res: list[list[int]]) -> None:
        if len(nums) == 0:
            return res.append(group.copy())
        
        for n in nums:
            group.append(n)
            self.helper(nums.difference([n]), group, res)
            group.pop()

if __name__ == '__main__':
    s = Solution()
    print(s.permute([1,2,3]))
    print(s.permute([0,1]))
    print(s.permute([1]))

File: leetcode/c++/46_permutations.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        helper(nums, 0, res);
        return res;
    }

    void helper(vector<int>& nums, int l, vector<vector<int>>& res) {
        if (l == nums.size()) {
            return res.push_back(nums);
        }

        for (int i = l; i < nums.size(); i++) {
            swap(nums[l], nums[i]);
            helper(nums, l + 1, res);
            swap(nums[l], nums[i]);
        }
    }
};